shrink

```diff
@@ -1,82 +1,269 @@
+			*) echo "Please answer yes or no.";;
+		esac
+	done
+	return $result
+}
+warn() {	# params: message
+    echo "$@" >&2
+}
+die() {	# params: exit code; message
+    rc="$1"
+    shift
+    warn "$@"
+    exit "$rc"
+}
+usage() {
+	msg="$*"
+	this=$(basename "$0")
+	if [ -n "$msg" ]; then
+		warn "$this: ERROR: $msg"
+	fi
+	warn "Usage: $this [-dk]"
+	warn '	Arguments:'
+	warn "		-c or --continue: Continue after failure. This restores all previously made *.sql.gz files too."
+	warn "		-d or --dry-run: Do not drop databases, remove files from $MYSQL_DATA_DIR, or restore databases."
+	warn '		-h or --help: Show this.'
+	if [ -n "$msg" ]; then
+		exit 1
+	fi
+	exit 0
+}
+
+
+# Command line arguments
+while :
+do
+	case "$1" in
+		-c) CONTINUE=1;;
+		--continue) CONTINUE=1;;
+		-d) DRYRUN=1;;
+		--dry-run) DRYRUN=1;;
+		-h) usage;;
+		--help) usage;;
+		-t) TEST=1;;	# secret option
+		#-o) shift; o="$1";;
+		#--) shift; break;;
+		-*) usage "bad argument $1";;
+		*) break;;
+	esac
+	shift
 done
-sudo service $MYSQL_SERVICE stop
-sudo rm -f /var/lib/mysql/ib*
-sudo service $MYSQL_SERVICE start
-if ! test $1; then
-  if [[ "$DBS" != "$SAVE_DBS" ]]; then
-    for DB in $DBS; do
-      if [ -f $DB.sql.gz ]; then
-        echo "Restoring database $DB..."
-        gunzip $DB.sql.gz
-        $MYSQL -u$MYSQL_USER -p$MYSQL_PASS -e "create database if not exists \`$DB\`"
-        $MYSQL -u$MYSQL_USER -p$MYSQL_PASS $DB < $DB.sql
-        if [ $? == 0 ]; then
-          echo "Database $DB restored successfully."
-          rm -f $DB.sql
-        else
-          echo "Restore of database $DB failed."
-        fi
-      fi
-    done
-  fi
-fi
-SIZE_AFTER=$(du /var/lib/mysql/ibdata1 | cut -d$'\t' -f1)
-SIZE_SAVED=$(($SIZE_BEFORE-$SIZE_AFTER))
-echo "Shrunk /var/lib/mysql/ibdata1 and recovered $SIZE_SAVED kilobytes."
+if [ -n "$DRYRUN" ]; then
+	echo '*** DRY RUN ***'
+fi
+
+
+# Make sure this script is running as root as sudo is not installed by default on Debian.
+if [ "$(id -u)" -ne 0 ]; then
+	die 1 'This script must be executed as root!';
+fi
+
+
+# Check that $MYSQL_DATA_DIR and the ibdata1 file exist.
+if [ ! -d "$MYSQL_DATA_DIR" ]; then
+	die 1 "Directory $MYSQL_DATA_DIR not found!"
+fi
+IBDATA1_FILE="$MYSQL_DATA_DIR/ibdata1"
+if [ ! -f "$IBDATA1_FILE" ]; then
+	die 1 "$IBDATA1_FILE not found!"
+fi
+
+
+# Check that required binaries are available.
+if [ -z "$(command -v mysql)" ]; then
+	die $? 'MySQL client (mysql) is not installed!'
+fi
+if [ -z "$(command -v mysqldump)" ]; then
+	die $? 'mysqldump not found!'
+fi
+GZIP="$(command -v pigz)"
+if [ -z "$GZIP" ]; then
+	GZIP="$(command -v gzip)"
+	if [ -z "$GZIP" ]; then
+		die $? 'gzip not found!'
+	fi
+fi
+
+
+# This line is the only bash dependency:
+if [ "$(basename "$SHELL")" = 'bash' ]; then
+	set -o pipefail
+fi
+
+
+# Make sure mysqld is running
+if [ -z "$(pidof mysqld)" ]; then
+	echo "Starting mysql service."
+	service mysql start || die 1 "Failed to start mysql: $?"
```

Comment:
This could be `service myslqd start` on some servers.  In fact, it might even need `systemctl start mysqld.service`.